Sources/Kontur.Rabbitmq/AsyncAmqpBasicConsumer.cs

```diff
@@ -32,47 +39,78 @@ class AsyncAmqpBasicConsumer<T> : IPublisher where T : class
             this.queue = queue;
         }
 
-        private void CancelConsuming(string consumerTag)
-        {
-            this.link.Dispose();
-            this.channel.BasicCancel(consumerTag);
-            this.channel.Close();
-            this.connection.Close();
-        }
-
         public IPublishingTag LinkTo(ITargetBlock<IMessage> target)
         {
-            if (this.connection != null || this.channel != null || this.convertBlock != null)
+            if (this.connection != null || this.channel != null || this.deserializeBlock != null)
             {
                 throw new InvalidOperationException("You could not link it more than once.");
             }
 
             this.connection = this.connectionFactory.CreateConnection();
             this.channel = this.connection.CreateModel();
 
-            this.convertBlock = new TransformBlock<AmqpMessage, IMessage>(
-                (Func<AmqpMessage, IMessage>) this.amqpMessageBuilder.Deserialize<T>);
+            this.deserializeBlock = new TransformBlock<AmqpDelivery, MessageResult>(
+                (Func<AmqpDelivery, MessageResult>)((delivery) =>
+               {
+                   try
+                   {
+                       return new MessageResult(
+                           this.amqpMessageBuilder.Deserialize<T>(delivery.Message));
+                   }
+                   catch (Exception ex)
+                   {
+                       return new MessageResult(
+                           new AmqpDeliveryError(delivery, ExceptionDispatchInfo.Capture(ex)));
+                   }
+               }));
+            this.unpackBlock = new TransformBlock<MessageResult, IMessage>(
+                (Func<MessageResult, IMessage>)(result => result.Value));
+            this.handleErrorBlock = new ActionBlock<MessageResult>(
+                ((Action<MessageResult>)(result =>
+                    {
+                        try
+                        {
+                            this.channel.BasicNack(result.Error.Delivery.DeliveryTag, false, false);
+                        }
+                        catch (Exception)
+                        {

```

Comment:
Empty catch?

Reply:
I created issue https://github.com/mzabolotko/Kontur/issues/48 to cover the common case of channel exception handling. This catch will be redisigned during its implementation.